### PR TITLE
[SMALLFIX] fixing a possible resource leak in AbstractClient

### PR DESCRIPTION
**Pull request by jsimsa**
in the previous logic, if `beforeDisconnect()` invokes any RPC, additional connection is be created and the first one is never be closed

### CODE REVIEW

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 1]
```diff
@@ -194,18 +194,13 @@ public synchronized void connect() throws IOException, ConnectionFailedException
    * if the client has not connected with the remote for a while, for example.
    */
   public synchronized void disconnect() {
-    if (mConnected) {
+    beforeDisconnect();
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/AbstractClient.java]: does it makes sense to call this and `afterDisconnect` even if `mConnected` is already false?

Comment (apc999) [inline core/common/src/main/java/alluxio/AbstractClient.java]: probably check `mConnected` in the beginning of this method

Comment (jsimsa) [inline core/common/src/main/java/alluxio/AbstractClient.java]: done

**Comment by apc999** [comment]
Jenkins, test this please

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 2]
```diff
@@ -196,14 +196,8 @@ public synchronized void connect() throws IOException, ConnectionFailedException
...
       LOG.debug("Disconnecting from the {} {} {}", getServiceName(), mMode, mAddress);
-      mConnected = false;
-    }
-    try {
       beforeDisconnect();
-      if (mProtocol != null) {
-        mProtocol.getTransport().close();
-      }
-    } finally {
+      mProtocol.getTransport().close();
```
Comment (apc999) [inline core/common/src/main/java/alluxio/AbstractClient.java]: don't we need to test `mProtocol != null`?

Comment (apc999) [inline core/common/src/main/java/alluxio/AbstractClient.java]: I read the source code, it seems  when `mConnected == true`, this variable will always be not null. discard my question.

**Comment by apc999** [comment]
+1 LGTM

File: core/common/src/main/java/alluxio/AbstractClient.java (hunk 2 above)
Comment (aaudiber) [inline core/common/src/main/java/alluxio/AbstractClient.java]: should we set mConnected = false?

Comment (apc999) [inline core/common/src/main/java/alluxio/AbstractClient.java]: yes
